src/Cassandra/TcpSocket.cs

```diff
@@ -389,39 +392,47 @@ protected virtual void OnClosing()
             _receiveBuffer = null;
         }
 
+        private void OnWriteFlushed()
+        {
+            var callback = Interlocked.Exchange(ref _writeFlushCallback, null);
+            if (callback != null)
+            {
+                callback();
+            }
+        }
+
         /// <summary>
         /// Sends data asynchronously
         /// </summary>
-        public virtual void Write(Stream stream)
+        public void Write(byte[] buffer, int length, Action onBufferFlush)
         {
-            //This can result in OOM
-            //A neat improvement would be to write this sync in small buffers when buffer.length > X
-            var buffer = Utils.ReadAllBytes(stream, 0);
+            Interlocked.Exchange(ref _writeFlushCallback, onBufferFlush);
             if (_isClosing)
             {
                 OnError(new SocketException((int)SocketError.Shutdown));
+                OnWriteFlushed();
                 return;
             }
             if (_sendSocketEvent != null)
             {
-                _sendSocketEvent.SetBuffer(buffer, 0, buffer.Length);
-                bool willRaiseEvent = false;
+                _sendSocketEvent.SetBuffer(buffer, 0, length);
```

Comment:
I doubt it is a good pattern to generate a large buffer and replace the `SocketAsyncEventArgs` buffer on every send. Instead, the outgoing stream could copied into the `SocketAsyncEventArgs` buffer using `Stream.Read`. Multiple calls to `Read` and `SendAsync` might be required to send the whole stream though.